### ISSUE TITLE
Allow no-restricted-imports to restrict sub-paths of imports

### ISSUE THREAD

**Issue by lencioni** [eslint/eslint#6963]
> The version of ESLint you are using.

3.3.1

> The problem you want to solve.

I have a package that contains some files that are considered to be part of its public API (e.g. "my-package/public/foo.js") which we consume by importing these files from the package directly (e.g. `import foo from 'my-package/public/foo.js';`). This package also contains files that are considered to be private and should not be imported (e.g. "my-package/private/bar.js"). I want to allow people to import the package (`import myPackage from 'my-package';`) or from the public directory, but not from the private directory.

> Your take on the correct solution to problem.

I originally considered [enhancing eslint-plugin-import's no-restricted-paths rule to allow for packages](https://github.com/benmosher/eslint-plugin-import/pull/500) but there @ljharb and @benmosher suggested enhancing the core no-restricted-imports rule instead.

**Comment by ljharb** [comment]
Since this would still just apply to absolute imports (ie, things in node_modules, not relative filesystem imports) it seems like it should be doable to support "startsWith" logic for anything - ie, `foo/bar/baz` could be prohibited by `foo/bar`.

**Comment by vitorbal** [comment]
I wonder if this wouldn't be better as a new option, though. If we add "startsWith" logic to the current implementation, then the behavior of the snippet below would change:

```javascript
/*eslint no-restricted-imports: ["error", "cluster"]*/
import cluster from 'cluster'; // fails
import foo from 'cluster/foo'; // does not fail currently, would start failing
```

**Comment by lencioni** [comment]
I'm happy with anything that supports my use-case.

One alternative to a new option would be allowing people to use `*` (or something like it) in the current option.

**Comment by nzakas** [comment]
Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

**Comment by ljharb** [comment]
@platinumazure @vitorbal @not-an-aardvark with the increased use of paths (`lodash/pick`, eg) in imports, is there any chance we could achieve a consensus such that this could be reopened?

**Comment by vitorbal** [comment]
@ljharb sorry, i completely lost track of this one. I'm more than happy to support it.
I just don't have the time to implement it myself right now, though. If someone is willing to whip up a PR then I can try to bug a few people so we can reach a consensus :)

**Comment by ljharb** [comment]
I'll happy to make a PR for a new option this weekend if you think that's enough to reopen the issue!

**Comment by ljharb** [comment]
As part of the PR, I'll also update `no-restricted-modules` in the same fashion. The three options:
 1. allow globs in the string (technically a breaking change, but unlikely)
 1. allow a single item to be either a string "foo", `{ path: "foo" }`, or `{ pattern: "foo/*" }`
 1. allow either a list of strings `"foo", "bar"`, or, `{ paths: ["foo", "bar"], patterns: ["foo/*"] }`

I'm leaning towards the last, which also paves the way for the string schema to be deprecated one day.

**Comment by ljharb** [comment]
I've put up #7433 in hopes that this can be reopened.

**Comment by platinumazure** [comment]
I think we can just accept the PR once we get enough support, so I don't think we need to reopen this specifically. Please track #7433 if you are interested in this feature.
